vdr/didx509/resolver.go

```diff
@@ -0,0 +1,316 @@
+	"net/url"
+	"slices"
+	"strings"
+)
+
+const (
+	MethodName                   = "x509"
+	X509CertChainHeader          = "x5c"
+	X509CertThumbprintHeader     = "x5t"
+	X509CertThumbprintS256Header = "x5t#S256"
+)
+
+type PolicyName string
+
+const (
+	PolicyNameSubject PolicyName = "subject"
+	PolicyNameSan     PolicyName = "san"
+)
+
+type SubjectPolicy string
+
+const (
+	SubjectPolicySerialNumber       SubjectPolicy = "serialNumber"
+	SubjectPolicyCommonName         SubjectPolicy = "CN"
+	SubjectPolicyLocality           SubjectPolicy = "L"
+	SubjectPolicyCountry            SubjectPolicy = "C"
+	SubjectPolicyOrganization       SubjectPolicy = "O"
+	SubjectPolicyOrganizationalUnit SubjectPolicy = "OU"
+	SubjectPolicyState              SubjectPolicy = "ST"
+	SubjectPolicyStreet             SubjectPolicy = "STREET"
+)
+
+type SanPolicy string
+
+const (
+	SanPolicyOtherName SanPolicy = "otherName"
+	SanPolicyDNS       SanPolicy = "dns"
+	SanPolicyEmail     SanPolicy = "email"
+	SanPolicyIPAddress SanPolicy = "ip"
+)
+
+var (
+	ErrDidMalformed         = errors.New("did:x509 is malformed")
+	ErrDidVersion           = errors.New("did:x509 does not have version 0")
+	ErrDidSanMalformed      = errors.New("did:x509 san policy is malformed")
+	ErrDidSubjectMalformed  = errors.New("did:x509 subject policy is malformed")
+	ErrUnkPolicyType        = errors.New("unknown policy type")
+	ErrUnkSubjectPolicyType = errors.New("unknown subject policy type")
+	ErrUnkSANPolicyType     = errors.New("unknown subject SAN type")
+	ErrX509ChainMissing     = errors.New("x509 rootCert chain is missing")
+)
+
+var _ resolver.DIDResolver = &Resolver{}
+
+// NewResolver creates a new Resolver.
+func NewResolver(pkiValidator pki.Validator) *Resolver {
+	return &Resolver{
+		pkiValidator: pkiValidator,
+	}
+}
+
+type Resolver struct {
+	pkiValidator pki.Validator
+}
+
+type X509DidReference struct {
+	Method      HashAlgorithm
+	RootCertRef string
+	PolicyName  PolicyName
+	PolicyValue string
+}
+
+// Resolve resolves a DID document given its identifier and corresponding metadata.
+func (r Resolver) Resolve(id did.DID, metadata *resolver.ResolveMetadata) (*did.Document, *resolver.DocumentMetadata, error) {
+	if id.Method != MethodName {
+		return nil, nil, fmt.Errorf("unsupported DID method: %s", id.Method)
+	}
+	ref, err := parseX509Did(id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	chainHeader, ok := metadata.GetProtectedHeaderChain(X509CertChainHeader)
+	if !ok {
+		return nil, nil, ErrX509ChainMissing
+	}
+	chain, err := parseChain(chainHeader)
+	if err != nil {
+		return nil, nil, err
+	}
+	_, err = findCertificateByHash(chain, ref.RootCertRef, ref.Method)
+	if err != nil {
+		return nil, nil, err
+	}
+	validationCert, err := findValidationCertificate(metadata, chain)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = validatePolicy(ref, validationCert)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = r.pkiValidator.Validate(chain)
```

Comment:
Discussed with @stevenvegt: No. this is not acceptable. The `pkiValidator` must use hardfail: return an error when it has no CRL cache for the chain, or when the cache is stale, i.e. older than `pki.maxupdatefailhours`. For other/existing use-cases the failure mode should still be configurable.

Reply:
@gerardsn @stevenvegt why is this different from the other cases where this is used (why hard-fail here?).

I find it dangerous to choose to hard-fail here, since it causes data exchanges to be blocked when the issuer of the UZI cert backing the credential's CRL is unreachable or expired. It's already quite a long chain, with more and more failure modes (we also have the revocation status of the credential itself to check), so hard-fail makes it less robust.
That's why we implemented soft-fail in the first place, so how is it different here?

Reply:
- before it was 'just' one of the layers of security. (Still requires did:nuts signatures to do anything)
- here it is the root of trust for your credentials. A CA should always be online, and even if it is offline for a few minutes, the cache is still available. Using softfail here means you accept credentials without checking the trust

I can make a PR that will make the required methods available

Reply:
@gerardsn I get what the general idea is, I just could use some help on how to implement this correctly. Should I create a new first class citizen for the pkiValidator instance with configuation elements, new defaults etc.? Or should I create my own instance with a specific configuration? Or should I clone / copy the TLS instance and just set the softfail = false?

Reply:
I will create a PR later today that you can use here.

Reply:
#3531 should give you `ValidateStrict` to use here

Reply:
What is procedure from here? Move on the this PR and add it later or wait for the https://github.com/nuts-foundation/nuts-node/pull/3531 to merge into main/master?

In the mean time I'm working on the latest issues.

Reply:
PR is merged so yo can use `ValidateStrict` now. I believe @reinkrul was still going to compare this with the spec, so maybe resolve all conflicts in the meantime so the tests can run again.

Reply:
The master + PR is merged into this PR, code is changed to `ValidateStrict`.